nodeJS/express/express_forms_validation.md

```diff
@@ -0,0 +1,565 @@
+    <form action="/users/<%= user.id %>/delete" method="POST" style="display:inline;">
+      <button type="submit" onclick="return confirm('Are you sure you want to delete this user?');">Delete</button>
+    </form>
+    <% }); %>
+  <% } %>
+</ul>
+```
+
+Then we add the logic to handle the request into our controller:
+
+```javascript
+// Tell the server to delete a matching user, if any. Otherwise, respond with an error.
+exports.usersDeletePost = asyncHandler(async (req, res) => {
+  const user = usersStorage.getUser(req.params.id);
+  if (!user) return res.status(404).send("User not found.");
+  usersStorage.deleteUser(req.params.id);
+  res.redirect("/");
+});
+```
+
+Don't forget to add the new route to your router!
+
+```javascript
+usersRouter.post("/:id/delete", usersController.usersDeletePost);
+```
+
+You'll notice we didn't need a `GET` route here since we're already redirecting to `/`.
+
+We could go much deeper into working safely with forms, but we'll stop there. By now you can already see how helpful `express-validator` is, and how you can do almost anything you want with the `req.body` object.
+
+To give you an overview of what this entire flow looks like visually, here's an example courtesy of MDN:
+
+![A diagram showing an outline of how a form interacts with a server using GET and POST requests.](https://developer.mozilla.org/en-US/docs/Learn/Server-side/Express_Nodejs/forms/web_server_form_handling.png)
+
+### Assignment
+
+<div class="lesson-content__panel" markdown="1">
+
+#### Add user details
+
+Expand the `User` model we created to include more details. Implement the following fields and validations to your model:
+
+- Email (required, must be formatted properly)
+- Age (optional, must be a number between 18 and 120)
+- Bio (optional, maximum 200 characters)
+
+Don't forget to update the view to display these new fields!
+
+#### Implement searching
+
+What if we want to search for a specific user in a list of thousands? We'll need a new route and view that lets clients search our list of users.
+
+1. Add a form (in `users.ejs` or another view) which accepts a `name` or `email` (or both!)
+1. Create a new route `/users/search` which accepts `GET` and `POST` requests.
+1. Add the search logic to your controller which searches your list for a matching user.
+   - Your `POST` request should handle searching for the user.
+   - Your `GET` request should then render the search result.
+1. Display the search results in a new view: `search.ejs`.
+
+</div>
+
+#### Further Reading
+
+- [OWASP.ORG](https://blog.presidentbeef.com/blog/2020/01/14/injection-prevention-sanitizing-vs-escaping/) will give you a good idea on what escaping means and how it can help further secure your web applications.
+- [express-validator](https://express-validator.github.io/docs/) contains the full documentation for `validator`, with some important sections being:
+  - [Getting Started](https://express-validator.github.io/docs/guides/getting-started)
```

Comment:
Nit for accessibility again, else navigating by link, a SR will just read "Getting Started - link"
```suggestion
  - [Getting Started with express-validator](https://express-validator.github.io/docs/guides/getting-started)
```